lib/segment/src/segment_constructor/segment_builder.rs

```diff
@@ -191,34 +229,113 @@ impl SegmentBuilder {
+        id_tracker.versions_flusher()()?;
+
         Ok(true)
     }
 
-    pub fn build(
-        mut self,
-        permit: CpuPermit,
-        stopped: &AtomicBool,
-    ) -> Result<Segment, OperationError> {
-        {
+    pub fn build(self, permit: CpuPermit, stopped: &AtomicBool) -> Result<Segment, OperationError> {
+        let (temp_path, destination_path) = {
+            let SegmentBuilder {
+                version,
+                id_tracker,
+                payload_storage,
+                mut vector_storages,
+                segment_config,
+                destination_path,
+                temp_path,
+                indexed_fields,
+            } = self;
+
+            let appendable_flag = segment_config.is_appendable();
+
+            payload_storage.flusher()()?;
+            let payload_storage_arc = Arc::new(AtomicRefCell::new(payload_storage));
+
+            id_tracker.mapping_flusher()()?;
+            id_tracker.versions_flusher()()?;
+            let id_tracker_arc = Arc::new(AtomicRefCell::new(id_tracker));
+
+            let payload_index_path = get_payload_index_path(temp_path.as_path());
+
+            let mut payload_index = StructPayloadIndex::open(
+                payload_storage_arc,
+                id_tracker_arc.clone(),
+                &payload_index_path,
+                appendable_flag,
+            )?;
+
+            for (field, payload_schema) in indexed_fields {
+                payload_index.set_indexed(&field, payload_schema)?;
+                check_process_stopped(stopped)?;
+            }
+
+            payload_index.flusher()()?;
+            let payload_index_arc = Arc::new(AtomicRefCell::new(payload_index));
+
             // Arc permit to share it with each vector store
             let permit = Arc::new(permit);
 
-            let mut segment = self.segment.take().ok_or(OperationError::service_error(
-                "Segment building error: created segment not found",
-            ))?;
+            let mut quantized_vectors = Self::update_quantization(
+                &segment_config,
+                &vector_storages,
+                temp_path.as_path(),
+                &permit,
+                stopped,
+            )?;
+
+            for (vector_name, vector_config) in &segment_config.vector_data {
+                let vector_index_path = get_vector_index_path(&temp_path, vector_name);
 
-            for (field, payload_schema) in &self.indexed_fields {
-                segment.create_field_index(segment.version(), field, Some(payload_schema))?;
-                check_process_stopped(stopped)?;
+                let Some(vector_storage) = vector_storages.remove(vector_name) else {
+                    return Err(OperationError::service_error(format!(
+                        "Vector storage for vector name {vector_name} not found on segment build"
+                    )));
+                };
+
+                vector_storage.flusher()()?;
```

Comment:
I propose to flush vector storages in `update_from` like for id tracker. because `update_from` is a place where vector storage is constructed. If you want to keep all flushes here in `build`, I propose to do it all in one place before dropping

Reply:
If we do it before dropping, we would need to go through Arc and RefCell stuff

Reply:
My idea was that flush should happen before we start using those objects for building index